### PR TITLE
Update Roslyn version

### PR DESCRIPTION
**Pull request by Youssef1313**
Commit-by-commit review is easier.

### CODE REVIEW

File: .editorconfig
Comment (mavasani) [inline .editorconfig:237]: Can we file a separate issue? Or can we wait for the new Microsoft.CodeAnalysis.Analyzers pre-release package with #6119 merged and update the repo to consume that before merging this PR?

Comment (Youssef1313) [reply .editorconfig:237]: I'm okay with waiting until #6119 is merged.

Comment (mavasani) [reply .editorconfig:237]: @Youssef1313 It is merged now. You can wait for new version showing up [here](https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet7/NuGet/Microsoft.CodeAnalysis.Analyzers/versions)

Comment (mavasani) [reply .editorconfig:237]: You'll have to update https://github.com/dotnet/roslyn-analyzers/blob/5463e16b5c0d1a83729babf5772de8b2917195c9/eng/Versions.props#L52

Comment (mavasani) [reply .editorconfig:237]: Or probably just bump https://github.com/dotnet/roslyn-analyzers/blob/5463e16b5c0d1a83729babf5772de8b2917195c9/eng/Versions.props#L49

**Comment by mavasani** [comment]
LGTM, though there seem to be many more violations to fix from the new package.

**Comment by Youssef1313** [comment]
@mavasani The last build had different failures between regular Ubuntu builds and source-build.

Looking at binlog of source-build:

image


Why is it still using an old version of the analyzer? 

Note BannedApiAnalyzers is using the correct version. Not sure what's special about Microsoft.CodeAnalysis.Analyzers

**Comment by Youssef1313** [comment]
I see now what's going on.

File: src/Directory.Build.props
Comment (Youssef1313) [inline src/Directory.Build.props:23]: We only include the following in non-source-build.

```xml
        <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="$(MicrosoftCodeAnalysisAnalyzersVersion)" />
```

So in source-build, we are supposed to not get the package. However, we get it as transitive dependency, and the transitive dependency is 3.0.0 which has the bug, causing source-build to fail.

Per `<!-- Disable code analysis for source build -->`, it seems we don't want to run analyzers for source-build. So I added `<RunAnalyzers>false</RunAnalyzers>`

File: src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/MicrosoftNetFrameworkAnalyzersResources.resx
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/MicrosoftNetFrameworkAnalyzersResources.resx:345]: This is passed only as a title, not a message. But I left the name as-is to minimize changes. Let me know if you want to rename this resource name.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: FYI @huoyaoyuan @buyaa-n

Comment (huoyaoyuan) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: Using 0U, 0L and 0UL may look better since there's no cast on int. Just stylish.

File: src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/MicrosoftNetFrameworkAnalyzersResources.resx
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/MicrosoftNetFrameworkAnalyzersResources.resx:345]: It’s fine to not change it in this PR

**Review by mavasani — APPROVED** [review]

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs
Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: We cannot use literal suffix for all numeric types, so the cast will remain in some and we'll end up with 3 styles:

- the plain literal (for `int`)
- the literal with suffix (for `long`, etc.)
- the cast (for `byte`, etc.)

I think it's less confusing to avoid suffix if the other styles will remain existing anyway.

Alternatively, I can keep the suppression and have the cast on all types.

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: I disabled auto merge until this conversation is resolved.

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: @huoyaoyuan Let me know if you still prefer using literal suffix, or if you think the current approach is good enough, or if you prefer keeping the cast and suppression.

Comment (huoyaoyuan) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: I'm really OK with either approach. For preference, I'm against suppression for stylish warnings.
I treat plain literal and literal with suffix as similar styles. This is my personal preference and we should follow the repo preference, if any.

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: > we should follow the repo preference, if any.

@mavasani Are there any coding guidelines regarding cast vs literal suffix? If not, I can push a commit to apply @huoyaoyuan's original suggestion.

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs:107]: Not really, I am fine with either approach here as well.

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSpanClearInsteadOfFill.cs]: @huoyaoyuan Can you take a look at the last commit? Thanks!

**Comment by Youssef1313** [comment]
@mavasani If the CI is green, hopefully this should be ready to merge.

**Review by mavasani — APPROVED** [review]

**Pull request merged**
